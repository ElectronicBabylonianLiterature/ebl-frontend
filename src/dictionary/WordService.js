class WordService {
  constructor(wordRepository) {
    this.wordRepository = wordRepository
  }

  find(id) {
    return this.wordRepository.find(id)
  }

  search(query) {
    return this.wordRepository.search(query)
  }

  update(word) {
    return this.wordRepository.update(word)
  }
}

export default WordService
