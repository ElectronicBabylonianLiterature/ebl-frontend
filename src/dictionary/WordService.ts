class WordService {
  private readonly wordRepository

  constructor(wordRepository) {
    this.wordRepository = wordRepository
  }

  find(id) {
    return this.wordRepository.find(id)
  }

  search(query) {
    return this.wordRepository.search(query)
  }

  update(word) {
    return this.wordRepository.update(word)
  }
}

export default WordService
