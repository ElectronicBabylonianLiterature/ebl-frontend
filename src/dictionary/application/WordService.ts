import Promise from 'bluebird'
import Word from 'dictionary/domain/Word'
import WordRepository from 'dictionary/infrastructure/WordRepository'
import _ from 'lodash'
import { stringify } from 'query-string'

export interface WordQuery {
  word?: string
  meaning?: string
  root?: string
  vowelClass?: string
}

class WordService {
  private readonly wordRepository: WordRepository

  constructor(wordRepository: WordRepository) {
    this.wordRepository = wordRepository
  }

  find(id: string): Promise<Word> {
    return this.wordRepository.find(id)
  }

  findAll(ids: string[]): Promise<readonly Word[]> {
    return this.wordRepository.findAll(ids)
  }

  search(query: WordQuery): Promise<Word[]> {
    return this.wordRepository.search(
      stringify(query, { skipEmptyString: true })
    )
  }

  searchLemma(lemma: string): Promise<readonly Word[]> {
    return _.isEmpty(lemma)
      ? Promise.resolve([])
      : this.wordRepository.searchLemma(lemma)
  }

  update(word: Word): Promise<Word> {
    return this.wordRepository.update(word)
  }
}

export default WordService
