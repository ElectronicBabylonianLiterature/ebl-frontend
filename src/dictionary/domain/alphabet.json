[
  "A",
  "a",
  "Ā",
  "ā",
  "Â",
  "â",
  "B",
  "b",
  "D",
  "d",
  "E",
  "e",
  "Ē",
  "ē",
  "Ê",
  "ê",
  "f",
  "G",
  "g",
  "Ĝ",
  "ĝ",
  "ŋ",
  "H",
  "h",
  "I",
  "i",
  "Ī",
  "ī",
  "Î",
  "î",
  "Y",
  "y",
  "K",
  "k",
  "L",
  "l",
  "M",
  "m",
  "N",
  "n",
  "P",
  "p",
  "Q",
  "q",
  "R",
  "r",
  "S",
  "s",
  "Ṣ",
  "ṣ",
  "Ś",
  "ś",
  "Š",
  "š",
  "T",
  "t",
  "Ṭ",
  "ṭ",
  "U",
  "u",
  "Ū",
  "ū",
  "Û",
  "û",
  "W",
  "w",
  "Z",
  "z",
  "₄",
  "½",
  "⅓",
  "¼",
  "⅙",
  "⅔",
  "⅚",
  "ø",
  " "
]
