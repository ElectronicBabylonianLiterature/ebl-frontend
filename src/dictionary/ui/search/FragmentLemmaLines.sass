.fragment-lines-with-lemma
    &__fragment-number
        vertical-align: top
        font-weight: normal
        padding: 1em .5em 0 0

        & ~ td
            padding-top: 1em
    &__line-number
        vertical-align: top
        text-align: right
        padding-right: .5rem
