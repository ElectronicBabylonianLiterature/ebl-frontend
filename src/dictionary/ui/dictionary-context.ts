import WordService from 'dictionary/application/WordService'
import React, { useContext } from 'react'

export const DictionaryContext = React.createContext<WordService | null>(null)

export function useDictionary(): WordService {
  const dictionary = useContext(DictionaryContext)
  if (dictionary === null) {
    throw new Error('useDictionary must be inside DictionaryContext.Provider')
  } else {
    return dictionary
  }
}
