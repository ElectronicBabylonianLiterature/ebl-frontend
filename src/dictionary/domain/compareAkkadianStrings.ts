import _ from 'lodash'
import alphabet from './alphabet.json'

const ignoredCharacters: readonly string[] = [
  'Ê¾',
  "'",
  ']',
  '?',
  '[',
  '-',
  ']',
  '.',
  'x',
  'X',
  'c',
  'C',
  'o',
  'O',
  'v',
]

const ignoredRegExp = new RegExp(
  _(ignoredCharacters).map(_.escapeRegExp).join('|'),
  'g'
)

function removeIgnoredCharacters(word: string): string {
  return word.replace(ignoredRegExp, '')
}

function checkForInvalidCharacters(...words: string[]): void {
  const invalidCharacters = words
    .flatMap((word) => word.split(''))
    .filter((character) => !alphabet.includes(character))
  if (!_.isEmpty(invalidCharacters)) {
    throw new Error(`Invalid character(s) ${invalidCharacters} in the input`)
  }
}

function compareAlphabet(word: string, anotherWord: string): number {
  return alphabet.indexOf(word) - alphabet.indexOf(anotherWord)
}

export function compareCleanedAkkadianString(
  word: string,
  anotherWord: string
): number {
  return compareAkkadianStrings(
    cleanAkkadianString(word),
    cleanAkkadianString(anotherWord)
  )
}

export default function compareAkkadianStrings(
  word: string,
  anotherWord: string
): number {
  const replacedWord = removeIgnoredCharacters(word)
  const anotherWordReplaced = removeIgnoredCharacters(anotherWord)
  checkForInvalidCharacters(replacedWord, anotherWordReplaced)

  return (
    _(replacedWord)
      .split('')
      .zipWith(anotherWordReplaced.split(''), compareAlphabet)
      .filter((result) => result !== 0)
      .map(Math.sign)
      .head() ?? Math.sign(replacedWord.length - anotherWordReplaced.length)
  )
}
export function cleanAkkadianString(akkadianString: string): string {
  return akkadianString
    .split('')
    .map((signChar) => (alphabet.indexOf(signChar) >= 0 ? signChar : ''))
    .join('')
}
