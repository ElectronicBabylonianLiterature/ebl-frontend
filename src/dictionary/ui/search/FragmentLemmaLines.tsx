import React from 'react'
import withData from 'http/withData'
import { QueryResult } from 'query/QueryResult'
import { Fragment } from 'fragmentarium/domain/fragment'
import FragmentService from 'fragmentarium/application/FragmentService'
import { LineColumns } from 'transliteration/ui/line-tokens'
import { createColumns } from 'transliteration/domain/columns'
import FragmentLink from 'fragmentarium/ui/FragmentLink'
import lineNumberToString from 'transliteration/domain/lineNumberToString'
import './FragmentLemmaLines.sass'
import _ from 'lodash'
import { TextLine } from 'transliteration/domain/text-line'
import { Col, Row } from 'react-bootstrap'
import LemmaQueryLink from '../display/LemmaQueryLink'

const linesToShow = 3

export function RenderFragmentLines({
  fragment,
  lemmaIds,
  linesToShow,
  totalLines,
}: {
  fragment: Fragment
  lemmaIds?: readonly string[]
  linesToShow: number
  totalLines: number
}): JSX.Element {
  const lines = fragment.text.lines.filter(
    (line) => line.type === 'TextLine'
  ) as TextLine[]

  return (
    <table>
      <tbody>
        {lines.map((line, index) => {
          const columns = [
            {
              span: 1,
              content: createColumns(line.content).flatMap(
                (column) => column.content
              ),
            },
          ]
          return (
            <tr key={index}>
              <td className={'fragment-lines-with-lemma__line-number'}>
                {lineNumberToString(line.lineNumber)}
              </td>
              <LineColumns
                columns={columns}
                maxColumns={1}
                highlightLemmas={lemmaIds || []}
              />
            </tr>
          )
        })}
        {totalLines > linesToShow && (
          <tr>
            <td></td>
            <td>And {totalLines - linesToShow} more</td>
          </tr>
        )}
      </tbody>
    </table>
  )
}

const FragmentLines = withData<
  { lemmaId: string; lineIndexes: readonly number[] },
  {
    museumNumber: string
    fragmentService: FragmentService
  },
  Fragment
>(
  ({ data: fragment, lemmaId, lineIndexes }): JSX.Element => (
    <RenderFragmentLines
      fragment={fragment}
      linesToShow={linesToShow}
      totalLines={lineIndexes.length}
      lemmaIds={[lemmaId]}
    />
  ),
  (props) =>
    props.fragmentService.find(
      props.museumNumber,
      _.take(props.lineIndexes, linesToShow)
    )
)

function FragmentLemmaLines({
  queryResult,
  fragmentService,
  lemmaId,
}: {
  queryResult: QueryResult
  fragmentService: FragmentService
  lemmaId: string
}): JSX.Element {
  return (
    <>
      {_.take(queryResult.items, 10).map((queryItem, index) => {
        return (
          <Row key={index}>
            <Col xs={1}>
              <FragmentLink number={queryItem.museumNumber}>
                {queryItem.museumNumber}
              </FragmentLink>
            </Col>
            <Col className={'fragmentlines-column'}>
              <FragmentLines
                lineIndexes={queryItem.matchingLines}
                museumNumber={queryItem.museumNumber}
                fragmentService={fragmentService}
                lemmaId={lemmaId}
                key={index}
              />
            </Col>
          </Row>
        )
      })}
    </>
  )
}

export default withData<
  {
    lemmaId: string
    fragmentService: FragmentService
  },
  unknown,
  QueryResult
>(
  ({ data: queryResult, fragmentService, lemmaId }): JSX.Element => {
    const total = queryResult.matchCountTotal.toLocaleString()
    return (
      <>
        <p>
          {total} matches&nbsp;
          <LemmaQueryLink lemmaId={lemmaId} />
        </p>
        <FragmentLemmaLines
          queryResult={queryResult}
          fragmentService={fragmentService}
          lemmaId={lemmaId}
        />
        <LemmaQueryLink lemmaId={lemmaId}>
          Show all {total} matches in Fragmentarium search&nbsp;
        </LemmaQueryLink>
      </>
    )
  },
  ({ fragmentService, lemmaId }) => fragmentService.query({ lemmas: lemmaId })
)
