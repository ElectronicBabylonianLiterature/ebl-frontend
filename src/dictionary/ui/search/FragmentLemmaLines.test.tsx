import FragmentService from 'fragmentarium/application/FragmentService'
import { QueryService } from 'query/QueryService'
import React from 'react'
import { act, render, screen } from '@testing-library/react'
import { dictionaryWord } from 'test-support/word-info-fixtures'
import FragmentLemmaLines from './FragmentLemmaLines'
import { fragment, lines } from 'test-support/test-fragment'
import { QueryItem, QueryResult } from 'query/QueryResult'
import Bluebird from 'bluebird'
import WordService from 'dictionary/application/WordService'
import { MemoryRouter } from 'react-router-dom'
import { DictionaryContext } from '../dictionary-context'
import { Fragment } from 'fragmentarium/domain/fragment'
import produce, { castDraft, Draft } from 'immer'
import { Text } from 'transliteration/domain/text'
import { TextLine, TextLineDto } from 'transliteration/domain/text-line'
import { atfTokenKur } from 'test-support/test-tokens'

jest.mock('fragmentarium/application/FragmentService')
jest.mock('dictionary/application/WordService')
jest.mock('query/QueryService')

const fragmentService = new (FragmentService as jest.Mock<
  jest.Mocked<FragmentService>
>)()
const wordService = new (WordService as jest.Mock<jest.Mocked<WordService>>)()
const queryService = new (QueryService as jest.Mock<
  jest.Mocked<QueryService>
>)()

const lineDto: TextLineDto = {
  ...lines[0],
  content: [{ ...atfTokenKur, uniqueLemma: [dictionaryWord._id] }],
}
const text = new Text({
  lines: [new TextLine(lineDto)],
})

const fragmentWithLemma = produce(fragment, (draft: Draft<Fragment>) => {
  draft.text = castDraft(text)
})

let container: HTMLElement

function renderFragmentLemmaLines() {
  container = render(
    <MemoryRouter>
      <DictionaryContext.Provider value={wordService}>
        <FragmentLemmaLines
          lemmaId={dictionaryWord._id}
          queryService={queryService}
          fragmentService={fragmentService}
        />
      </DictionaryContext.Provider>
    </MemoryRouter>
  ).container
}

describe('Show Fragmentarium entries', () => {
  beforeEach(async () => {
    const queryItem: QueryItem = {
      museumNumber: {
        prefix: 'Test',
        number: 'Fragment',
        suffix: '',
      },
      matchingLines: [0],
      matchCount: 1,
    }
    const queryResult: QueryResult = { items: [queryItem], matchCountTotal: 1 }
    queryService.query.mockReturnValue(Bluebird.resolve(queryResult))
    fragmentService.find.mockReturnValue(Bluebird.resolve(fragmentWithLemma))

    await act(async () => renderFragmentLemmaLines())

    expect(queryService.query).toBeCalledWith(dictionaryWord._id)
  })

  it('shows the fragment number', async () => {
    expect(screen.getByText(fragmentWithLemma.number)).toBeVisible()
  })

  it('shows the matching Fragmentarium line', async () => {
    expect(container).toMatchSnapshot()
  })
})
