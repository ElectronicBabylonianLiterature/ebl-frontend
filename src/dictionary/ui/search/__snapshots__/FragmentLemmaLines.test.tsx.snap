// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Show Fragmentarium entries shows the matching Fragmentarium line 1`] = `
<div>
  <p>
    1
     attestations
  </p>
  <div
    class="row"
  >
    <div
      class="col-1"
    >
      <a
        href="/fragmentarium/Test.Fragment"
      >
        Test.Fragment
      </a>
    </div>
    <div
      class="fragmentlines-column col"
    >
      <table>
        <tbody>
          <tr>
            <td
              class="fragment-lines-with-lemma__line-number"
            >
              10
            </td>
            <td
              colspan="1"
            >
              <span
                class="Transliteration__wordSeparator Transliteration__wordSeparator--AKKADIAN"
              >
                 
              </span>
              <span
                class="Transliteration__Word Transliteration__Word--AKKADIAN Transliteration__Word--highlight"
              >
                <span
                  class="word-info__trigger"
                  role="button"
                >
                  <span
                    class=""
                  >
                    <span
                      class="Transliteration__Reading"
                    >
                      <span
                        class=""
                      >
                        <span
                          class="Transliteration__ValueToken"
                        >
                          <span
                            class=""
                          >
                            kur
                          </span>
                        </span>
                        <sup
                          class="Transliteration__modifier"
                        />
                        <sup
                          class="Transliteration__flag"
                        />
                      </span>
                    </span>
                  </span>
                </span>
              </span>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
